compiler/src/dotty/tools/dotc/transform/sjs/PrepJSInterop.scala

```diff
@@ -279,6 +279,39 @@ class PrepJSInterop extends MacroTransform with IdentityDenotTransformer { thisP
           val ctorOf = ref(jsdefn.JSPackage_constructorOf).appliedToTypeTree(tpeArg)
           ref(jsdefn.Runtime_newConstructorTag).appliedToType(tpeArg.tpe).appliedTo(ctorOf)
 
+        /* Rewrite js.dynamicImport[T](body) into
+         *
+         * runtime.dynamicImport[T](
+         *   new DynamicImportThunk { def apply(): Any = body }
+         * )
+         */
+        case Apply(TypeApply(fun, List(tpeArg)), List(body))
+            if fun.symbol == jsdefn.JSPackage_dynamicImport =>
+          val span = tree.span
+          val currentOwner = ctx.owner
+
+          assert(currentOwner.isTerm, s"unexpected owner: $currentOwner at ${tree.sourcePos}")
+
+          val cls = newNormalizedClassSymbol(currentOwner, tpnme.ANON_CLASS, Synthetic | Final,
+              List(jsdefn.DynamicImportThunkType), coord = span)
+          val constr = newConstructor(cls, Synthetic, Nil, Nil).entered
```

Comment:
I kind of wanted to use `tpe.AnonClass` here, but it only creates *forward* methods, whereas I need a new method (`apply`) that actually contains the tree `body`. Is there a better way to do this?

Reply:
Indeed, there was some opportunity for refactoring. In fact, it extended to two other places in the codebase which followed the same pattern. I have added a commit that factors out the common pattern from those 4 places. Does it look good to you?